File: js/src/compat.js
// Compatibility folding — NFKC, where the same text wears different costumes.
//
// Failure mode #9 (nfkc-compatibility-fold): Japanese input carries several
// width variants of the same letters. Halfwidth katakana (ﾊﾝｶｸ, U+FF61..U+FF9F)
// and fullwidth ASCII (Ａ１２３, U+FF01..U+FF5E) look like — and mean — the same
// thing as ハンカク and A123, but they are entirely different code points. A
// search, a dedup, a username check, or a payment-form validator that compares
// raw, or even after canonical (NFC) normalization, treats them as unequal: the
// record is missed, the duplicate is kept, the "already taken" check passes.
//
// This is distinct from NFC/NFD (#7). That fixture is about *canonical*
// equivalence (é vs e+◌́, which NFC settles). Width variants are only
// *compatibility* equivalent — NFC leaves them apart on purpose, and only the
// compatibility forms (NFKC / NFKD) fold them together. The fix is to NFKC
// before comparing.

/** Compatibility-composed form: folds width variants together. */
export function nfkc(s) {
  return s.normalize("NFKC");
}

/** Buggy reference: raw comparison — width variants look unequal. */
export function naiveEquals(a, b) {
  return a === b;
}

/**
 * Insufficient reference: canonical (NFC) comparison. It settles é vs e+◌́ but
 * deliberately leaves halfwidth/fullwidth width variants apart, so this still
 * reports ﾊﾝｶｸ and ハンカク as different. Shown to prove #7's fix is not enough here.
 */
export function canonicalEquals(a, b) {
  return a.normalize("NFC") === b.normalize("NFC");
}

/** Correct: compatibility (NFKC) comparison folds the width variants. */
export function compatEquals(a, b) {
  return a.normalize("NFKC") === b.normalize("NFKC");
}

File: go/compat.go
package cjkfixtures

// Compatibility folding (failure mode #9, nfkc-compatibility-fold). Halfwidth
// katakana (ﾊﾝｶｸ) and fullwidth ASCII (Ａ１) are different code points from their
// normal-width twins (ハンカク, A1) but mean the same thing. A raw comparison, and
// even a canonical (NFC, failure mode #7) comparison, reports them unequal; only
// compatibility normalization (NFKC) folds them.
//
// The Go standard library has no normalizer (NFKC lives in golang.org/x/text,
// an external module). To stay dependency-free, compatFold below applies a
// minimal compatibility fold: fullwidth ASCII by arithmetic (U+FF01..U+FF5E ->
// U+0021..U+007E) plus a small table of the halfwidth katakana these fixtures
// use. It covers the width variants in the tests, not the full NFKC mapping.

// Halfwidth katakana -> fullwidth katakana, for the characters the fixtures use.
var halfwidthKatakana = map[rune]rune{
	0xff76: 0x30ab, // ｶ -> カ
	0xff78: 0x30af, // ｸ -> ク
	0xff8a: 0x30cf, // ﾊ -> ハ
	0xff9d: 0x30f3, // ﾝ -> ン
}

// compatFold returns a compatibility-folded key. Two width variants of the same
// text produce the same key.
func compatFold(s string) string {
	out := make([]rune, 0, len(s))
	for _, r := range s {
		switch {
		case r >= 0xff01 && r <= 0xff5e: // fullwidth ASCII -> ASCII
			out = append(out, r-0xfee0)
		default:
			if fw, ok := halfwidthKatakana[r]; ok {
				out = append(out, fw)
			} else {
				out = append(out, r)
			}
		}
	}
	return string(out)
}

// CompatFold returns the compatibility-folded (NFKC-style) form of a string.
func CompatFold(s string) string {
	return compatFold(s)
}

// CompatEquals is correct: compare compatibility folds so width variants match.
func CompatEquals(a, b string) bool {
	return compatFold(a) == compatFold(b)
}
